tempesta_fw/http.c

```diff
@@ -4141,25 +4148,33 @@ tfw_http_msg_process(void *conn, TfwFsmData *data)
 {
 	int r = T_OK;
 	TfwStream *stream = &((TfwConn *)conn)->stream;
-	unsigned int trail = data->trail;
+	unsigned int data_left = data->len;
 	struct sk_buff *next;
 
 	if (data->skb->prev)
 		data->skb->prev->next = NULL;
-	for (next = data->skb->next; data->skb;
+	for (next = data->skb->next; data->skb && data_left > 0;
 	     data->skb = next, next = next ? next->next : NULL, data->off = 0)
 	{
 		if (likely(r == T_OK || r == T_POSTPONE)) {
 			data->skb->next = data->skb->prev = NULL;
-			data->trail = !next ? trail : 0;
+			data->len = min_t(unsigned int, data_left,
+					  data->skb->len - data->off);
+			data_left -= data->len;
 			r = TFW_CONN_H2(conn)
 				? tfw_h2_frame_process(conn, data)
 				: tfw_http_msg_process_generic(conn, stream, data);
 		} else {
-			kfree(data->skb);
+			kfree_skb(data->skb);
 		}
 	}
 
+	while (data->skb) {
+		next = data->skb->next;
+		kfree_skb(data->skb);
+		data->skb = next;
+	}
```

Comment:
I was puzzled with the addition for some time, so it'd be reader friendly to have a comment here. We need the code for `!data_left && data->skb` and we reach the condition only for unusable for HTTP data, TLS tag in particular. I was puzzled because the `for` loop is complex enough and from quick reading it looks like the upper layer must set `data->skb = NULL` or isn't in charge to free skb, but it's not true...